# Changelog

## [1.2.0](https://github.com/3d-omics/mg_assembly/compare/v1.1.0...v1.2.0) (2024-11-01)


### Features

* assemble quast + assemble report ([ad3f2f8](https://github.com/3d-omics/mg_assembly/commit/ad3f2f83a3b2f38037aed277428f8fb33f87192c))
* insert mg_preprocess, delete redundant code, cram -&gt; bam ([6589dc1](https://github.com/3d-omics/mg_assembly/commit/6589dc1c5f904ae84b00be761220cf9898190d46))
* multiqc report for viruses ([b6581b7](https://github.com/3d-omics/mg_assembly/commit/b6581b7f7a71dd9979e94f4a4a5e5f5baf162647))
* split environments ([36f3e8a](https://github.com/3d-omics/mg_assembly/commit/36f3e8a16ece0af1074a64027cf189d0ca91a591))


### Bug Fixes

* add parallel to envs ([394d345](https://github.com/3d-omics/mg_assembly/commit/394d345c076d4c8ca01909d39a5f507dccf3b0f9))
* add pigz to envs ([a642477](https://github.com/3d-omics/mg_assembly/commit/a64247750018e08e9732af3c3cd68e04d6800393))
* add rename to environment ([12a50b2](https://github.com/3d-omics/mg_assembly/commit/12a50b2c5e6530d2b20618d77d92431f8e351cea))
* **dram:** params -&gt; output ([6d082b4](https://github.com/3d-omics/mg_assembly/commit/6d082b41359fbc10f080b1f9653c4b67a56b4f4d))
* **drep:** make report parsing a localrule ([0f15e8d](https://github.com/3d-omics/mg_assembly/commit/0f15e8d4346ea9691cef39ba6de9b1664dd3a11d))
* **drep:** rename rule in profile and reduce time and ram ([be19aac](https://github.com/3d-omics/mg_assembly/commit/be19aacbeb7214641e35053b585222d36793d1d7))
* **profile:** double on failure in bins, name report rules properly ([389650c](https://github.com/3d-omics/mg_assembly/commit/389650cd7b52d14ab1b11af99ed53971f68b7c5d))
* txt -&gt; tsv ([1903ad7](https://github.com/3d-omics/mg_assembly/commit/1903ad74e5e42261a11d16cbb56649c55ed55016))

## [1.1.0](https://github.com/3d-omics/mg_assembly/compare/v1.0.0...v1.1.0) (2024-09-26)


### Features

* add nonpareil curves to multiqc ([bc140b1](https://github.com/3d-omics/mg_assembly/commit/bc140b1f5d770e55b3e879972519820084bec71e))
* add per secondary ani bowtie2 reporst ([7ba9276](https://github.com/3d-omics/mg_assembly/commit/7ba9276bb1e16d63cbe26a3e68bfa7d40e8a0951))
* add preprocess__samtools ([f07f298](https://github.com/3d-omics/mg_assembly/commit/f07f298f10dd1e87e052b02a23fa128fb00fd421))
* do prokaryote annotation as in mg_annotate ([3f4a78c](https://github.com/3d-omics/mg_assembly/commit/3f4a78c261877441541c7f862f8ca60066943c0b))
* do prokaryote quantification over all anis ([fd8decd](https://github.com/3d-omics/mg_assembly/commit/fd8decdae81323b0fb578768a9fa6fca152a7fab))
* extract drep params to config file ([64bb9f2](https://github.com/3d-omics/mg_assembly/commit/64bb9f2d4eb4b717e1f126a1286fbf9601c847e4))
* harmonize with mg_quant; handle crams instead of pairs of fastq files ([7257054](https://github.com/3d-omics/mg_assembly/commit/7257054165a531b1f170866fabadc844ad42dd38))
* pin dependencies ([38169a6](https://github.com/3d-omics/mg_assembly/commit/38169a660016a74183538a37a3d528f97fc62a56))
* stop storing u1 and u2 from fastp ([648f7c4](https://github.com/3d-omics/mg_assembly/commit/648f7c4ca4b311d6818d0dce15af8f55f346170e))
* update multiqc to 1.24 ([d9fc566](https://github.com/3d-omics/mg_assembly/commit/d9fc5669818bc5f2cb6f95cdfb91cd2e5be3d7f6))
* update nonpareil and multiqc so they do the jsoning and the reporting ([5f73a0a](https://github.com/3d-omics/mg_assembly/commit/5f73a0a22df60506e5114218d8d5e7fb4d3c9a95))
* update readme ([fe9b0f6](https://github.com/3d-omics/mg_assembly/commit/fe9b0f6788fccf4e20b7ade48d0480bd3728a9bb))
* update rulegraphs ([454e777](https://github.com/3d-omics/mg_assembly/commit/454e777ab54b4eadaf1bc7f51cf7f59f30a0e09e))


### Bug Fixes

* correct folder ([5196257](https://github.com/3d-omics/mg_assembly/commit/519625730481a68e08a1e97b459f0c122aa0cdf8))
* correct nonpareil ([4c40ed2](https://github.com/3d-omics/mg_assembly/commit/4c40ed21b3705df84246458bd7690d93923af116))
* correct singlem rule names and fix them in the profile ([3aeb0b7](https://github.com/3d-omics/mg_assembly/commit/3aeb0b7589da14d22afcb01095228dcba14cddf7))
* harmonization ([1b218bc](https://github.com/3d-omics/mg_assembly/commit/1b218bcb3e7e96be00af53dcaa028b36db689e5c))
* nonpareil ([685f75c](https://github.com/3d-omics/mg_assembly/commit/685f75c98e4ad6ba4c727e4d330d75319f7ba72e))


### Performance Improvements

* disable groups ([77d906a](https://github.com/3d-omics/mg_assembly/commit/77d906a9a81e69078c2b45fd71290ff65e5497fa))
* fix kraken2 rule name ([90b4f27](https://github.com/3d-omics/mg_assembly/commit/90b4f270507949988b01c0e4feef553f9bfcbdff))
* run read symlinking locally ([1ad21cd](https://github.com/3d-omics/mg_assembly/commit/1ad21cd6a70fa7e2c93bf4cdb60dc034f18fc0a8))

## 1.0.0 (2024-06-13)


### Features

* add CI badge ([cfc54e8](https://github.com/3d-omics/mg_assembly/commit/cfc54e8287d8341d19f89cb2bc1742f5084fb766))
* add paths to the three reference genomes that we will always use ([d50fd2c](https://github.com/3d-omics/mg_assembly/commit/d50fd2cb1e19ed194bafe3b06ffc075c68318297))
* add references and indexes to cache ([e854001](https://github.com/3d-omics/mg_assembly/commit/e854001557be79a05b35933fe3f57be5aa147de0))
* **annotate:** remove code to dowload databases ([7d3a3cd](https://github.com/3d-omics/mg_assembly/commit/7d3a3cda7ddadc118cf5f02e6fe951c56f3bd4d9))
* **annotate:** remove empty folder ([c6aa727](https://github.com/3d-omics/mg_assembly/commit/c6aa7274672fabc43fe1afb8b337181924017659))
* **annotate:** remove env ([923eaad](https://github.com/3d-omics/mg_assembly/commit/923eaadc68c721903a673d6dbe684669f013e76c))
* **annotate:** run everything by default ([370c9f4](https://github.com/3d-omics/mg_assembly/commit/370c9f4cebfe754a951df36876baddce73edfe28))
* **annotate:** separate dereplicate and annotate ([1fb84df](https://github.com/3d-omics/mg_assembly/commit/1fb84df439176e1cad71565362f8edf8568c5ed2))
* **annotate:** simplify env ([e91ffbc](https://github.com/3d-omics/mg_assembly/commit/e91ffbcb3087a68dd077edaea5e9dac2a4d7eb91))
* **assemble/bowtie2:** write index within the rule ([5a43b9a](https://github.com/3d-omics/mg_assembly/commit/5a43b9a10deadb7cd41d6e1b21d05436fce008a9))
* **assemble/metabat2:** move minifunction to lambda ([15ec930](https://github.com/3d-omics/mg_assembly/commit/15ec93066022fa20e79227cb15927337e0deea93))
* **assemble/metabat2:** move minifunction to lambda ([5a9d1c3](https://github.com/3d-omics/mg_assembly/commit/5a9d1c3c0dc636cea00163e0f129cffa3f0fee38))
* **assemble:** insert drep env to main env ([9065862](https://github.com/3d-omics/mg_assembly/commit/9065862187ccc08de8a5c35bb710adc9f33cffb2))
* **assemble:** insert magscot env to main env ([9deb3dc](https://github.com/3d-omics/mg_assembly/commit/9deb3dcf1b156813b780bc7123e6a72d83731bd9))
* **assemble:** insert maxbin2 env to main env ([83dccd3](https://github.com/3d-omics/mg_assembly/commit/83dccd3afc475b9ae67e22ca8a632ff881e3396f))
* **assemble:** insert metabat2 env to main env ([69bd15b](https://github.com/3d-omics/mg_assembly/commit/69bd15b3be4123d0166cb2f4ddc404044303932b))
* **assemble:** put drep inside assemble ([0e15318](https://github.com/3d-omics/mg_assembly/commit/0e153188c1b7fef5c1099dd87844e4dbbbfecdc5))
* **assemble:** refactor ([3089791](https://github.com/3d-omics/mg_assembly/commit/308979112b6f7a101d5735a39931ebd8f8b96acd))
* **assemble:** refactor and delete dead code for binners ([610d7ea](https://github.com/3d-omics/mg_assembly/commit/610d7eab33c1815e4d50415221bf1fafbfd66275))
* **assemble:** refactor megahit functions ([25ba606](https://github.com/3d-omics/mg_assembly/commit/25ba606fc04f6c37ba13a30b07cf40c56fb20aca))
* **assemble:** rename rules ([1cef682](https://github.com/3d-omics/mg_assembly/commit/1cef682637a365058ddf338e6dd4c26a32dadc7a))
* **assemble:** reorder subworkflows ([844e3fa](https://github.com/3d-omics/mg_assembly/commit/844e3fa0ce751f4ee48d75e5c8bbc8e787d89fb0))
* **assemble:** sort environment ([6f72c73](https://github.com/3d-omics/mg_assembly/commit/6f72c73f791fdcfc410c692cb55a2f0ba7fdac13))
* **assembly:** process gzipped files, remove crap ([5d74b66](https://github.com/3d-omics/mg_assembly/commit/5d74b66fb839b52f95f3a102db241716fa601391))
* **bin:** rename rules ([e6ba83f](https://github.com/3d-omics/mg_assembly/commit/e6ba83ff836102a83555d14244070a4782b69723))
* **concoct:** remove dead code ([7056b18](https://github.com/3d-omics/mg_assembly/commit/7056b1805a7a91a07817206f93e023e6183a0bd7))
* **dereplicate:** rename rules ([8a45e95](https://github.com/3d-omics/mg_assembly/commit/8a45e950da8e8899f59e926082fa2092c9e6e7b8))
* **dram:** disable database download ([a263b36](https://github.com/3d-omics/mg_assembly/commit/a263b3657f701ab7781c258a719727aff2ecee49))
* **dram:** remove everything in case of failure and move log ([bd68340](https://github.com/3d-omics/mg_assembly/commit/bd683401495057911cf5042c4f8eecb4d9df5cea))
* **drep:** bgzip fasta files. Accept bgzipped files downstream ([b36d4a4](https://github.com/3d-omics/mg_assembly/commit/b36d4a48ed1d27d8c4de54c02e6769a5b999fb5e))
* **drep:** move log in case of failure ([7122d6c](https://github.com/3d-omics/mg_assembly/commit/7122d6c5fa9f90e5736f055cc417ff06c664f055))
* **drep:** remove output files and folder on startup, since they are still there on a repeated run ([a8b82d0](https://github.com/3d-omics/mg_assembly/commit/a8b82d0cd95232d6bab691d16fe37b45eb760341))
* **drep:** tarballs as output ([b01000f](https://github.com/3d-omics/mg_assembly/commit/b01000fd251f5ee522f369643d0c4448450eb12b))
* **dunders:** convert main, envs and functions into dunder files ([b9f4b5b](https://github.com/3d-omics/mg_assembly/commit/b9f4b5ba12cffbd3fa115721d3aa4e40c32ebcfa))
* exapnd annotate/genomad output files ([edef9ae](https://github.com/3d-omics/mg_assembly/commit/edef9ae796bc84c853c4a0927965e0c1d24091df))
* extract all resources and threads to profile ([b6fc200](https://github.com/3d-omics/mg_assembly/commit/b6fc2002ba2d7e54f2f2aeea21c76037587bbc8b))
* extract all resources and threads to profile ([ffab624](https://github.com/3d-omics/mg_assembly/commit/ffab624d557c8c0d7651485cb18181baafa2375d))
* failed log processing ([ae01c0d](https://github.com/3d-omics/mg_assembly/commit/ae01c0db132a01c9594a4ff0882b27c6c5208f9d))
* failed log processing ([ae01c0d](https://github.com/3d-omics/mg_assembly/commit/ae01c0db132a01c9594a4ff0882b27c6c5208f9d))
* **fastp:** compress with bgzip and level 9 ([83d9f69](https://github.com/3d-omics/mg_assembly/commit/83d9f69acddd2de1c69ca02539ed1bb03a9b17e3))
* **fastp:** use 8GB ([71870bb](https://github.com/3d-omics/mg_assembly/commit/71870bb2b5421f93c7e5be8a2befc1a7743d89e6))
* **features:** put the commented references in the usual order ([4bc67b8](https://github.com/3d-omics/mg_assembly/commit/4bc67b8a7f24d48025a770a05fa1ca2232b5bd79))
* **features:** sort databases alphabetically ([9b5fab2](https://github.com/3d-omics/mg_assembly/commit/9b5fab2f36ea8c0c341a6558589062ecb920ed8c))
* **functions:** refactor and use under notation ([4aa3e67](https://github.com/3d-omics/mg_assembly/commit/4aa3e679d7535779a487adc4f8e83b15ea89e714))
* general cleaning ([329e101](https://github.com/3d-omics/mg_assembly/commit/329e1013b2a11e5945fef0b551f7ea77ae561e33))
* **genomad:** add keras ([4310c38](https://github.com/3d-omics/mg_assembly/commit/4310c38c312aedda394b4a1660145fa2af820626))
* **gha/annotate:** only quast ([1df663c](https://github.com/3d-omics/mg_assembly/commit/1df663c1db8cec7c638482722838302bdf5a8ed4))
* **gha:** download environments only when necessary ([d58f37b](https://github.com/3d-omics/mg_assembly/commit/d58f37b981112b2ba4aab497751a03eea2ce5e1c))
* **gha:** uncomment caching to see what happens ([909ef65](https://github.com/3d-omics/mg_assembly/commit/909ef65b9b96ce71fc06cbcf6c8fc9bc0a8e4811))
* group databases ([d49ac01](https://github.com/3d-omics/mg_assembly/commit/d49ac0109e74d25dfcd832a048088e7085e4420e))
* group jobs by sample or assembly ([43a9daf](https://github.com/3d-omics/mg_assembly/commit/43a9dafa9c845a432e11aa94688f34b6f4b7900f))
* **gtdbtk:** clean before execution and move log ([acc3eb2](https://github.com/3d-omics/mg_assembly/commit/acc3eb229122970452baf78f788f66713387117d))
* **gtdbtk:** extract final trees before tarballing everything ([8a550ac](https://github.com/3d-omics/mg_assembly/commit/8a550ac1c3a6108ced1167a9b064e8c4aff79d30))
* **gtdbtk:** process log retries ([bdfab74](https://github.com/3d-omics/mg_assembly/commit/bdfab7457a418fff4691c233a177fd569e8124ca))
* **gtdbtk:** tarballs as output ([d169afb](https://github.com/3d-omics/mg_assembly/commit/d169afba14d54d7dacce090f6f6230c648d00be6))
* **helpers:** rename rules ([8520490](https://github.com/3d-omics/mg_assembly/commit/852049067569180d51ec930d26e7d769bd01fc75))
* ignore reports ([8ef6549](https://github.com/3d-omics/mg_assembly/commit/8ef654998e9a2ca9fc347ec5044870176ae11800))
* improve nonpareil rules ([a15ed7f](https://github.com/3d-omics/mg_assembly/commit/a15ed7fb777192993d2c86760c9fa97d7d95d9f8))
* **kraken2:** print date in log ([c249009](https://github.com/3d-omics/mg_assembly/commit/c249009aa46f811ffa0a308a1ab52ef2a68e20cf))
* **kraken2:** process samples in order ([d511d78](https://github.com/3d-omics/mg_assembly/commit/d511d781c46f3eed4c44654ba416378dbd061ec6))
* **kraken2:** use a for loop, print the hostname, linting ([8ea7fbc](https://github.com/3d-omics/mg_assembly/commit/8ea7fbcfc1693e93c72b357d45c6eb9f09a3766a))
* linting ([ff3e573](https://github.com/3d-omics/mg_assembly/commit/ff3e57370d8e59a37e4951edd5c3e057b1d1c2d5))
* linting ([30a37e2](https://github.com/3d-omics/mg_assembly/commit/30a37e243a043a43e776182cec02ea684c647d6a))
* **logging:** process bowtie2 failed logs ([9f12010](https://github.com/3d-omics/mg_assembly/commit/9f12010e4ed5a1e0eaf2f73e686c254f1b64f2e4))
* **logging:** process bowtie2 failed logs ([8403f56](https://github.com/3d-omics/mg_assembly/commit/8403f56dc960f91f1f7fc1cc4ae30ba9b252270b))
* **logging:** process drep failed logs. clean dram and gtdbtk ([cc1a156](https://github.com/3d-omics/mg_assembly/commit/cc1a156efa5fd6bce5e713e37b05f1f286e6962f))
* **logging:** process megahit failed logs ([fff1127](https://github.com/3d-omics/mg_assembly/commit/fff1127b8435de8489b0ee8fa180e85b3d14fd87))
* **logging:** process report failed logs ([5ce5e54](https://github.com/3d-omics/mg_assembly/commit/5ce5e542bfd907c5cbe1d7f83c9315e12a145885))
* **logging:** refactor ([1c8067e](https://github.com/3d-omics/mg_assembly/commit/1c8067e96a23689bece695d15bb9d160fa02f4f5))
* **magscot:** process log retries ([927dfd4](https://github.com/3d-omics/mg_assembly/commit/927dfd4d82ce627261a37b42901555bb9377b048))
* **magscot:** remove dead code ([1535a81](https://github.com/3d-omics/mg_assembly/commit/1535a81bd7aceeca782935ecec79fac10718e574))
* **magscot:** remove dead code ([428be0c](https://github.com/3d-omics/mg_assembly/commit/428be0ccefab763751704976e83a73c537571144))
* **magscot:** work with gzipped files, remove unnecesary ones ([dfacb7d](https://github.com/3d-omics/mg_assembly/commit/dfacb7dc3535892c6bcf65ad62c6aa1b6ba36ac4))
* **maxbin2:** solve possible escape sequence ([30fdda0](https://github.com/3d-omics/mg_assembly/commit/30fdda06cb3d190c624f2f2f9f40ff7f5da1647f))
* **maxbin:** clean up folder ([0ce6484](https://github.com/3d-omics/mg_assembly/commit/0ce6484eeafefbbc702f097179087c1499368ba8))
* merge assemble and bin ([13a8514](https://github.com/3d-omics/mg_assembly/commit/13a8514bd3eb41ed62c92da33191b8dddb1bcdb4))
* move test to .github ([ab901bc](https://github.com/3d-omics/mg_assembly/commit/ab901bc52d2d2e35064b6cd3a62a6c13c4122438))
* **nonpareil:** decompress fastq, process, delete ([66ca2fa](https://github.com/3d-omics/mg_assembly/commit/66ca2fa3a593c39d4fdb5d520d196163371c283b))
* pack databases as in mjolnir ([63bbab5](https://github.com/3d-omics/mg_assembly/commit/63bbab50dc9c370e39002b7d1e0519a92f437f19))
* **params:** reduce splits ([dab64ec](https://github.com/3d-omics/mg_assembly/commit/dab64ec76508b7e3d4bb89d5db675d33d7e96c06))
* **pre-commit:** use R linters. lint all files, exclude magscot for obvious reasons ([71ae479](https://github.com/3d-omics/mg_assembly/commit/71ae4790992f10ce313846f6a557a74e93488867))
* **preprocess/functions:** delete dead code ([d17529c](https://github.com/3d-omics/mg_assembly/commit/d17529c80816a9379fb27cd66253c85a025d4be3))
* **preprocess/kraken2:** move minifunctions to lambdas ([26fcd34](https://github.com/3d-omics/mg_assembly/commit/26fcd34f703376895898d01350b241e473292c63))
* **preprocess/nonpareil:** move minifunction to lambda ([07b5356](https://github.com/3d-omics/mg_assembly/commit/07b53565bca513b70e70f51bb4eeea4e36ac5fb2))
* **preprocess:** extract resources to profile ([28b3488](https://github.com/3d-omics/mg_assembly/commit/28b3488172a594d1ebd0be71dcedbb942b969092))
* **preprocess:** linting ([999b4c2](https://github.com/3d-omics/mg_assembly/commit/999b4c2bccfae4fd2bb96db705b0f6ad6e662d2f))
* **preprocess:** make singlem microbial fraction ([a9f7a51](https://github.com/3d-omics/mg_assembly/commit/a9f7a518ee84a23958c47f5ce777bd5d0c4031e8))
* **preprocess:** refactor adapter retrieval. don't assume nextera ([b8f677b](https://github.com/3d-omics/mg_assembly/commit/b8f677bd817dc30dffd99d813dfb274e3ba3a999))
* **preprocess:** refactor compose_rg_extra ([c3bca61](https://github.com/3d-omics/mg_assembly/commit/c3bca61b808f69deeefb06b11cf3889d252b00d1))
* **preprocess:** refactor input files for bowtie2 ([f7bcfac](https://github.com/3d-omics/mg_assembly/commit/f7bcfac411fae813f117dfbb74320d97a28e096b))
* **preprocess:** refactor input files for bowtie2 ([02412aa](https://github.com/3d-omics/mg_assembly/commit/02412aaf19d67f3e2f7fff55ee8e0eb7e05436c0))
* **preprocess:** regroup functions ([d235523](https://github.com/3d-omics/mg_assembly/commit/d235523854d55c14c5a31472ca891e26e7db4ba3))
* **preprocess:** rename rules ([86ffb8d](https://github.com/3d-omics/mg_assembly/commit/86ffb8d1694655da3276cc7c9b60ea1f7392507b))
* **preprocess:** reorganize default rules. Run all by default, run some in the tests ([999521b](https://github.com/3d-omics/mg_assembly/commit/999521b1e5c49747694aa38b9a92b1caf55b070c))
* **quantify/coverm:** refactor functions, accept cram as input ([a402a0c](https://github.com/3d-omics/mg_assembly/commit/a402a0cf0561424e60372aad7604377670d4f8ac))
* **quantify:** extract quantification rules ([38a1f87](https://github.com/3d-omics/mg_assembly/commit/38a1f87bca1b1c0aeda8595ec2fa7edfbb43c6cd))
* **reads:** refactor functions ([c5b9cdd](https://github.com/3d-omics/mg_assembly/commit/c5b9cdd8577e6ac920f43df6b388cbbbadbd6b7f))
* **reads:** rename rules ([986765c](https://github.com/3d-omics/mg_assembly/commit/986765ce3d23e42c4aed00d78378144be28cd4ec))
* **reads:** run linking remotely ([664e3f6](https://github.com/3d-omics/mg_assembly/commit/664e3f6510b73aa4cb5413bc2fbfc86118ac54d5))
* **reads:** split into link and fastqc ([f06c01c](https://github.com/3d-omics/mg_assembly/commit/f06c01cb35fc8b9b8a318f576b563ce261e8abab))
* reassign databases ([59855a0](https://github.com/3d-omics/mg_assembly/commit/59855a04c80052c1a3081f06f15ef4872345f97a))
* **reference:** rename rules ([7095183](https://github.com/3d-omics/mg_assembly/commit/70951838a04f2b894054481348e3dfa7970badcb))
* **reference:** split references between hosts and mags for consistency with other pipelines ([cd69f5b](https://github.com/3d-omics/mg_assembly/commit/cd69f5bb75b4a78f12a3c470e296b79aea915eab))
* remove coverm from preprocess ([15a581a](https://github.com/3d-omics/mg_assembly/commit/15a581a8238530b9bf8812509de332f9543be78a))
* remove dead code ([254a3eb](https://github.com/3d-omics/mg_assembly/commit/254a3eb3b9e61789c337450d45fce0251a0625e4))
* remove empty line ([8bbfa53](https://github.com/3d-omics/mg_assembly/commit/8bbfa532b46cac742c6f065a060b27ccc2435b90))
* remove memory from params.yml ([2c67e40](https://github.com/3d-omics/mg_assembly/commit/2c67e408485a8266f0b9c684f9373d1fa730d1d7))
* remove virify ([6f825eb](https://github.com/3d-omics/mg_assembly/commit/6f825eb324f027d83e8e6d41039f03bd01c084cf))
* **reorganization:** __env__.yml =&gt; __environment__.yml ([c2f7610](https://github.com/3d-omics/mg_assembly/commit/c2f7610ee5c4cfeb10dd82c02fcdca9770821430))
* reorganize viral part ([fa421da](https://github.com/3d-omics/mg_assembly/commit/fa421daa710b244f889068f9a34b15fe48be7173))
* **report:** add assemble step report ([f8a01d6](https://github.com/3d-omics/mg_assembly/commit/f8a01d6dd5575a504b51615ed4d114547c136ee0))
* **report:** add the quantify report ([c31fd4d](https://github.com/3d-omics/mg_assembly/commit/c31fd4d8a9cffea14188dd1eeb8d34d0ac1a41a6))
* **report:** clean unsused functions ([ec1b3c9](https://github.com/3d-omics/mg_assembly/commit/ec1b3c97b93640e582e279215bcf3af388fb0089))
* **report:** create per-sample reports ([fa0eb86](https://github.com/3d-omics/mg_assembly/commit/fa0eb86d741854467f9167a70473824f35a27516))
* **report:** disable assembly step report - too many files to process ([e25a572](https://github.com/3d-omics/mg_assembly/commit/e25a57221868e40848a187eec5c80355967d6c2a))
* **report:** remove assembly report. it can become huge ([a77a736](https://github.com/3d-omics/mg_assembly/commit/a77a736d9fb8d74ecc4791297c227c05dfe3df57))
* **report:** rename rules ([aa53c2e](https://github.com/3d-omics/mg_assembly/commit/aa53c2e182afab28b9f04841488ccaf6c96741c4))
* **report:** update multiqc ([cef31fa](https://github.com/3d-omics/mg_assembly/commit/cef31facb39876f1d9e18ef6fa27721c5785acdc))
* restructure params.yml. remove all extras ([e47c2ef](https://github.com/3d-omics/mg_assembly/commit/e47c2eff3a6568f72048533e649ea76f6820c285))
* script to pack results and reports ([46a2d8a](https://github.com/3d-omics/mg_assembly/commit/46a2d8a3444460c4dde21275b6a4b6212e5ee65f))
* shadow ([d78b136](https://github.com/3d-omics/mg_assembly/commit/d78b136e21c4f05a8165aef792d59967c8676132))
* simplify test ([19eb10c](https://github.com/3d-omics/mg_assembly/commit/19eb10c87b233d1dcf46136c79e46a0c6acd180c))
* **singlem:** update to 0.16.0 ([26a1c73](https://github.com/3d-omics/mg_assembly/commit/26a1c735c08f9d7f3de23b7fb9cbc3d3627309f5))
* **test:** test, but skip drep ([2593f64](https://github.com/3d-omics/mg_assembly/commit/2593f64301f3deb510d1d084ee27f23a7173c8ba))
* **test:** update testing to new structure ([9425488](https://github.com/3d-omics/mg_assembly/commit/9425488c4cb4739f2238818f79976c958882e804))
* update and ignore virsorter2 mock db ([642d252](https://github.com/3d-omics/mg_assembly/commit/642d252160f5a2557564b1e437329b69215043b1))
* update environments ([cac5693](https://github.com/3d-omics/mg_assembly/commit/cac5693e781027743a06891ca51728d076186720))
* Update figure ([cc7c448](https://github.com/3d-omics/mg_assembly/commit/cc7c4482cf53068ba34ea4920e2e75b6b6924c4d))
* update genomad mock db ([00c56b9](https://github.com/3d-omics/mg_assembly/commit/00c56b9a95ad08f4706c63599aa0e7888deba455))
* update genomda db ([1a4011e](https://github.com/3d-omics/mg_assembly/commit/1a4011e70b22018d7df3558ed410a631b1101f54))
* update hooks ([cd4e7c4](https://github.com/3d-omics/mg_assembly/commit/cd4e7c43abd64743a6cbbbe3ec544ff5b85380b4))
* update readme ([40fc5d5](https://github.com/3d-omics/mg_assembly/commit/40fc5d52f442b46542f0151bd0c0f692f0988422))
* update readme ([15f170c](https://github.com/3d-omics/mg_assembly/commit/15f170ccd285e7ac5353f6e769b71dfd5b9e72c5))
* update rulegraphs ([194d30e](https://github.com/3d-omics/mg_assembly/commit/194d30e6b0cd816bcc30ed0657b8b6b58fdbc22a))
* **viral:** add checkv + filtering step ([96d6a49](https://github.com/3d-omics/mg_assembly/commit/96d6a4910c09748aaef3dce3b8d9ee538464155e))
* **viral:** add dram-v ([78d7b84](https://github.com/3d-omics/mg_assembly/commit/78d7b844e6f316f8fbde284164273442fed77dda))
* **viral:** add genomad ([4820e14](https://github.com/3d-omics/mg_assembly/commit/4820e1430ac5694954c7888c9076826334cc8ad8))
* **viral:** add htseq ([b625077](https://github.com/3d-omics/mg_assembly/commit/b62507780329046f897ab370fdb7ab6a15c5b123))
* **viral:** add mmseqs ([6435508](https://github.com/3d-omics/mg_assembly/commit/6435508202ea4a97957517d8bb71d38a8a12df8f))
* **viral:** add mmseqs. refactor ([bfa9414](https://github.com/3d-omics/mg_assembly/commit/bfa9414f75d1932f7b232ce5c621eb010a0b9252))
* **viral:** added dedupe. fixed dramv ([3521c62](https://github.com/3d-omics/mg_assembly/commit/3521c6204a098e80d0ce0d4bbed8612c28d72ab5))
* **viral:** added diamond ([73ff4d4](https://github.com/3d-omics/mg_assembly/commit/73ff4d43704f85c64c9e3485109e7db24e4c2331))
* **viral:** get the gene- and contig-wise gff files from genomad ([ee10444](https://github.com/3d-omics/mg_assembly/commit/ee10444eaa54f391f7c5f89d8b4f2361a544f62b))
* **viral:** linting ([246aada](https://github.com/3d-omics/mg_assembly/commit/246aada606e55c2fb8681e167f57b9c08ccc969d))
* **viral:** snakefmt ([34bc772](https://github.com/3d-omics/mg_assembly/commit/34bc7729b7bbf7b38ab7ae829c6a1add968b9e1d))
* **viral:** switch annotation from checkv to genomad ([f23758d](https://github.com/3d-omics/mg_assembly/commit/f23758d63d36775576f0d2a6ef1221fa6936455a))
* **viral:** virsorter ([cfafe43](https://github.com/3d-omics/mg_assembly/commit/cfafe4360850b1565787734dc8e73c607ea2b5c8))
* **virify:** make rules nonlocal ([a263976](https://github.com/3d-omics/mg_assembly/commit/a263976290e995aa1200c0458e295870402f70f8))
* **virify:** rename rules ([52d38bf](https://github.com/3d-omics/mg_assembly/commit/52d38bfd0b262f2582e68256272f0a09fd3d2cbe))
* **viruses:** added checkv ([a9bbab9](https://github.com/3d-omics/mg_assembly/commit/a9bbab98d08f1f441ade2bca6e478f9ab91a5986))


### Bug Fixes

* add prokaryotes__annotate to its correct folder ([1af0636](https://github.com/3d-omics/mg_assembly/commit/1af0636575eabb72b474c57eedcc7add1631af3e))
* add seqtk ([416c869](https://github.com/3d-omics/mg_assembly/commit/416c8698712687806324bbe1c05403d884f9aeb5))
* annotate/genomad output dir ([69cc1b9](https://github.com/3d-omics/mg_assembly/commit/69cc1b9aef780a48b9b27459461cc7eb80cf624b))
* **annotate:** point gtdbtk to the correct environment ([be5adb5](https://github.com/3d-omics/mg_assembly/commit/be5adb55c972efa76ade0d5e61c029a39a9e7793))
* **annotate:** raise dram mem to 32GB ([6822283](https://github.com/3d-omics/mg_assembly/commit/68222837a6982ff16638dacdaf4136da99284fb2))
* ar53 ([4c8e882](https://github.com/3d-omics/mg_assembly/commit/4c8e882b0431290fc91601b3bd3e3f710293c2cd))
* **assemble/bowtie2:** raise ram to 16gb [skip ci] ([615c0b5](https://github.com/3d-omics/mg_assembly/commit/615c0b51b03c6bdf6481656ba1f8c03ee6288c1d))
* **assemble:** cram to bam has to use 8GB not (MB ([f61308f](https://github.com/3d-omics/mg_assembly/commit/f61308f6edb2aec68b4443100f153a9c8a9f719e))
* **bowtie2:** CRAM to BAM in a single thread ([7cdec6b](https://github.com/3d-omics/mg_assembly/commit/7cdec6bdcaa7bc3d776321e1b15d54fd7c5ae416))
* compress with wildcards instead of using find ([b2025e2](https://github.com/3d-omics/mg_assembly/commit/b2025e266feba4b38fa2444d559a576a67138b1f))
* compress/decompress for drep ([c95f2dc](https://github.com/3d-omics/mg_assembly/commit/c95f2dca1d327e9d2025e9173d9abdf67e38e9a5))
* **concoct:** add pigz ([94fb4f6](https://github.com/3d-omics/mg_assembly/commit/94fb4f6a48171010a1ad72b93d97a9f392884b54))
* **concoct:** add threads ([6602595](https://github.com/3d-omics/mg_assembly/commit/66025952fed48516fe765fcb9b0bc29aa8df1623))
* **concoct:** snakefmt ([5f7349c](https://github.com/3d-omics/mg_assembly/commit/5f7349c87f9679eca6be7b4e53224196b4697d2d))
* disable gpu in gnomad, but enable nn ([4cf4d6d](https://github.com/3d-omics/mg_assembly/commit/4cf4d6d3b411e1aeca0193045dae913bc5db81fe))
* do not test dereplicate: gha runs out of disk space ([7433184](https://github.com/3d-omics/mg_assembly/commit/74331842d012e01ddeb3fa7daccbcc5da2e4ba12))
* dont create tarball of dereplicated genomes ([6b3d09c](https://github.com/3d-omics/mg_assembly/commit/6b3d09c178ecd1808bdec0ac6eb9130785a8369a))
* drep can't handle gzipped files ([bb16380](https://github.com/3d-omics/mg_assembly/commit/bb163807cf74269fbb4799b614278a5f6c46b455))
* **drep:** change flags in bgzip that are not in gzip ([294aa7f](https://github.com/3d-omics/mg_assembly/commit/294aa7f1fcdcd4610ecdc01343046a4d398a6f21))
* **drep:** compress correctly with bgzip ([6de3b30](https://github.com/3d-omics/mg_assembly/commit/6de3b30504ad293826634485ec6f089876b079f8))
* **drep:** compress fasta files ([5b557f7](https://github.com/3d-omics/mg_assembly/commit/5b557f73c99644d40c6dc20fc7bba60ebd72b26c))
* **drep:** it is --compress-level. test only annotate__quast ([b7ef6c8](https://github.com/3d-omics/mg_assembly/commit/b7ef6c867edd6de07473b940a0626ed2c4498799))
* **drep:** typo ([ccf71c7](https://github.com/3d-omics/mg_assembly/commit/ccf71c7759682b452e844378e2076d734a9cb5ae))
* extra ` ([ff91b15](https://github.com/3d-omics/mg_assembly/commit/ff91b15ee884375298145f22ed4af21c45a77f2b))
* **features:** put the pig genome in the correct place ([1f031cc](https://github.com/3d-omics/mg_assembly/commit/1f031cce5ed98c8af41a904bd2573aa41afc6939))
* fix tar errors ([dfc6ed2](https://github.com/3d-omics/mg_assembly/commit/dfc6ed2ff5c1243912c4e85864a6b0a8fdf7feea))
* fix test script ([cd2024c](https://github.com/3d-omics/mg_assembly/commit/cd2024c875dd7e9b14505d9ec25b8b6a6b2ab867))
* **functions:** move around and simplify functions ([53c8a9b](https://github.com/3d-omics/mg_assembly/commit/53c8a9bb21da696601372d72435c9888d15c5e50))
* **genomad:** disable nn classification by default ([9ca36f4](https://github.com/3d-omics/mg_assembly/commit/9ca36f4dd03ba0e20d1d3c6f52693c1b88785362))
* **genomad:** enable neural network ([4f304dd](https://github.com/3d-omics/mg_assembly/commit/4f304ddf5f450eaadc248343838b316f755f2bb5))
* **genomad:** fix output path ([e3a53cc](https://github.com/3d-omics/mg_assembly/commit/e3a53ccab1dc3113a1ad4121dcbed2905d71b57f))
* **gha:** cache with custom action ([5f52f2d](https://github.com/3d-omics/mg_assembly/commit/5f52f2d90739d557f1d88a15fb34655c25b3213d))
* **gha:** cp instead of gzipping ([443eb2a](https://github.com/3d-omics/mg_assembly/commit/443eb2aa6a0b085599c2bf10ccd8adc173de47d5))
* **gha:** Disable cache ([76a7f81](https://github.com/3d-omics/mg_assembly/commit/76a7f8131d7c38f029ded65b8e729c0a1d06b4ab))
* **gha:** fix CI with the new rules ([43f1b6c](https://github.com/3d-omics/mg_assembly/commit/43f1b6c7c29fd66ee4b5b0b2e2b54a5a1e44ea4e))
* **gha:** point to the temporary gha for snakemake ([4e6befd](https://github.com/3d-omics/mg_assembly/commit/4e6befdb2a083c3549c10628c596572db264111e))
* **gha:** pre -&gt; preprocess ([a4bb04b](https://github.com/3d-omics/mg_assembly/commit/a4bb04bbd6968835dfc79a98b3781befaec420c0))
* **gha:** pre -&gt; preprocess again ([5b89900](https://github.com/3d-omics/mg_assembly/commit/5b899000f2fd5e307605a0c30f30fc62e07d1db0))
* **gha:** sudo touch ([37b5810](https://github.com/3d-omics/mg_assembly/commit/37b581001e30af900f0641dbe432d0c5d1d68d73))
* **gha:** update path of files ([ff671ae](https://github.com/3d-omics/mg_assembly/commit/ff671aeaff318b35fbbfd8ef7d58415e3a25d5a7))
* **gha:** update pre-commit hooks ([3809f7b](https://github.com/3d-omics/mg_assembly/commit/3809f7b92584db771ee7c4459ba1d74a33310920))
* **gha:** update steps ([5b8a488](https://github.com/3d-omics/mg_assembly/commit/5b8a4880a107403b487821fd64918b5ca6a9c2b6))
* **gha:** use sudo ([c584c04](https://github.com/3d-omics/mg_assembly/commit/c584c0486360dd9746e0914244a1e3e30572248f))
* **gtdbtk dram:** typo ([7813dc5](https://github.com/3d-omics/mg_assembly/commit/7813dc5a17debec92066708ac5f96bede06cd939))
* **gtdbtk:** forgot to rename environment ([f581ab1](https://github.com/3d-omics/mg_assembly/commit/f581ab1b14090b6629b152a766ded1fd36487b10))
* **gtdbtk:** raise RAM ([557dc73](https://github.com/3d-omics/mg_assembly/commit/557dc732fba4a64e9afdc62d6a5fb1868c71c077))
* **gtdbtk:** reintroduce retries ([e254ba6](https://github.com/3d-omics/mg_assembly/commit/e254ba6d5dde9c32ffe8531b5beffdf9aa9f137a))
* **gtdbtk:** remove cleaning step ([130fb4e](https://github.com/3d-omics/mg_assembly/commit/130fb4e359c705ec5bfbe473a43b0ea7fd20a696))
* **gtdbtk:** typo in cp ([4c93354](https://github.com/3d-omics/mg_assembly/commit/4c93354d318a656ec76763cc3d68c0f40013d030))
* gzip drep inputs and outputs ([4c81eae](https://github.com/3d-omics/mg_assembly/commit/4c81eae0d10dca7caa7138ab7c81b31b5f1aced6))
* **htseq:** naming ([20f35cf](https://github.com/3d-omics/mg_assembly/commit/20f35cf29316dbbaa4a5018b848233c109cf2d49))
* import rules with correct name ([7906dab](https://github.com/3d-omics/mg_assembly/commit/7906dab515a6869b3f692e2627580a5983a2e7e7))
* **kraken2:** chown and chmod just in case ([8742b7c](https://github.com/3d-omics/mg_assembly/commit/8742b7c9e108e7084ae59d592d8ec78647063a3c))
* linting ([05048a3](https://github.com/3d-omics/mg_assembly/commit/05048a31d0d9b31c36a4599e564405c3be06d2a4))
* linting ([c59813d](https://github.com/3d-omics/mg_assembly/commit/c59813d38f479ecde1b9fa87e54ec8868b0c7112))
* linting ([202e23f](https://github.com/3d-omics/mg_assembly/commit/202e23f6d15ea44765495a2475b3eeff817c2afa))
* linting ([724c724](https://github.com/3d-omics/mg_assembly/commit/724c724964f5efc0d2b3e7720751d135da5d127a))
* pin smafa to the version that can work in the database ([304e93f](https://github.com/3d-omics/mg_assembly/commit/304e93fda439a8f1e7fe5add9ec4c7cbb0751912))
* **pre-commit:** disable R linting ([bf6aa7e](https://github.com/3d-omics/mg_assembly/commit/bf6aa7e54f4eb2285911499ece477a0f4483f1d0))
* **pre-commit:** missing linted files ([3dd3bf4](https://github.com/3d-omics/mg_assembly/commit/3dd3bf43e1d0dcecba09554fee1f940371c3aa77))
* **pre-commit:** update hooks ([70e0093](https://github.com/3d-omics/mg_assembly/commit/70e0093ccc6db9ce29aa0a390fa246d9251ff488))
* precommit files ([ea09c36](https://github.com/3d-omics/mg_assembly/commit/ea09c361765ee7c4c8298d0572f8b80371b4a597))
* **preprocess/samtools:** remove idxstats ([dc906bf](https://github.com/3d-omics/mg_assembly/commit/dc906bf4dd5ed2d9e1a4c410b722be60ce9c3c0f))
* **preprocess/samtools:** remove idxstats ([158f496](https://github.com/3d-omics/mg_assembly/commit/158f49679cfd9f20d5316ea2ab948080c9c9e77e))
* **preprocess:** handle empty hosts ([e46f8ef](https://github.com/3d-omics/mg_assembly/commit/e46f8ef8e0eb9479197a836a9e65144fe21cf459))
* **preprocess:** raise runtime to singlem pipe ([950336b](https://github.com/3d-omics/mg_assembly/commit/950336b6803ef79f411cc39b392356a7c1dfae63))
* **preprocess:** rename kraken2 job ([6bd5fc3](https://github.com/3d-omics/mg_assembly/commit/6bd5fc3d9f047572d60e06504e70a879cb9d1e08))
* put fq files in the correct place. linting ([06ac0a2](https://github.com/3d-omics/mg_assembly/commit/06ac0a27592612586d9a151fefdc218d1bc0291e))
* put k2 database in the exact folder as in mjolnir ([c10bd0f](https://github.com/3d-omics/mg_assembly/commit/c10bd0f601b03d4ceaa34fdc17f10be80e39fd5f))
* **readme:** fix badge ([a92f196](https://github.com/3d-omics/mg_assembly/commit/a92f196393edaff6f8ce7b731a6203a26ae34669))
* **reads:** raise the time and memory of links becayse mjolnir complains about both ([643186e](https://github.com/3d-omics/mg_assembly/commit/643186eea2d7c521dd051b1ca72c62c344a3721f))
* redirect logs to the correct file ([271badf](https://github.com/3d-omics/mg_assembly/commit/271badf4dc1a1d94a13a9deeee0bc64bd9c0ac27))
* **reference:** wrong env ([6eaa2a0](https://github.com/3d-omics/mg_assembly/commit/6eaa2a052b008e53ca14ff4c79edd67218856f49))
* remove unnecessary programs ([b66a629](https://github.com/3d-omics/mg_assembly/commit/b66a629744aa8f37eb40284c0946ed55ca72b3ba))
* report -&gt; reports ([1d9a6d2](https://github.com/3d-omics/mg_assembly/commit/1d9a6d2b8f188bfccaf154f89322f4c96284b536))
* **report:** double ram in preprocess report ([6273960](https://github.com/3d-omics/mg_assembly/commit/62739600e6e65e80d9c40f82bb184ea5087a9845))
* **report:** introduce retries to preprocess ([f6aa0f6](https://github.com/3d-omics/mg_assembly/commit/f6aa0f639790100f7bcc84e241f5d82bcba5c30e))
* **report:** raise runtime in preprocess ([6629662](https://github.com/3d-omics/mg_assembly/commit/66296620ce86949be3395ad041bede76a84a41e3))
* reports are slurm jobs now ([50c9fa3](https://github.com/3d-omics/mg_assembly/commit/50c9fa356e0e3aa277d67675dfeb34056bcf804b))
* **rulegraph:** convert tabs to spaces ([ff7b621](https://github.com/3d-omics/mg_assembly/commit/ff7b621091b8d86705c42a9a930b595bda7c83bd))
* run until dram-v ([f001779](https://github.com/3d-omics/mg_assembly/commit/f001779acbc905fc161c529e9dcce0b4993e4f49))
* **samtools:** do the crais and reports properly ([796cb42](https://github.com/3d-omics/mg_assembly/commit/796cb4226dbd3462c813027b8a5f0af05aff3137))
* **samtools:** restore crai generation ([96af236](https://github.com/3d-omics/mg_assembly/commit/96af2366bf0fdd6b96998838d51f176462bc3f13))
* **singlem:** bug ([7d04504](https://github.com/3d-omics/mg_assembly/commit/7d045040ca1048373315a2daddefc19306cc8874))
* **singlem:** hardcoded file ([78a3f3e](https://github.com/3d-omics/mg_assembly/commit/78a3f3e1a6eafa3b26a755b513a95175d25be3ec))
* **singlem:** raise mem to 8GB ([8fdf715](https://github.com/3d-omics/mg_assembly/commit/8fdf715f6c889fb66db1426d1b010722474e85bd))
* **singlem:** raise runtime ([bee23b6](https://github.com/3d-omics/mg_assembly/commit/bee23b6c3f5c623104d150cbcc206a7d36fad3fc))
* **singlem:** raise singlem pipe runtime ([666375a](https://github.com/3d-omics/mg_assembly/commit/666375a5b213875c7121dd00b67a06c12c28b250))
* **singlem:** use 1 CPU and 16GB of RAM ([84052e5](https://github.com/3d-omics/mg_assembly/commit/84052e58da9c4d74e2a8fd2bab25ea5646a37716))
* **snakefile:** handle samples not used in any assembly ([858d5af](https://github.com/3d-omics/mg_assembly/commit/858d5afbfe4b07ee46016ff2195635250fadaee9))
* **snakefile:** handle samples not used in any assembly, part 2 ([a99d12c](https://github.com/3d-omics/mg_assembly/commit/a99d12c9d09c0e19f25a2f757a1b92a596cf783c))
* **snakefile:** handle samples not used in any assembly, part 3 ([d4d60d6](https://github.com/3d-omics/mg_assembly/commit/d4d60d6749ebad56175bf0e724bd7b5414c4d7e3))
* snakelint ([e12bc9a](https://github.com/3d-omics/mg_assembly/commit/e12bc9a2aa029921368b358a64a0fb04006bb64a))
* snakelint ([01c305a](https://github.com/3d-omics/mg_assembly/commit/01c305a65b0cadf0f849ecee3604303f06b0ea2a))
* **test:** incorrect rule name ([79f50a7](https://github.com/3d-omics/mg_assembly/commit/79f50a7cbfbc0ea79a9468bbd339725219e449fe))
* typo in drep, don't use shadow in dramv ([8f2d336](https://github.com/3d-omics/mg_assembly/commit/8f2d3362da5a76d639dafd9d0933db3f9985189e))
* untangle bacterial and viral quantifications ([fd64423](https://github.com/3d-omics/mg_assembly/commit/fd64423765976a22d415f67e3c0c5f8c015259d0))
* update test with new rule names ([f5c344d](https://github.com/3d-omics/mg_assembly/commit/f5c344dc6fa8fe7130a881b27f2133f2e736114c))
* viral/quantify paths for reference ([db6eb89](https://github.com/3d-omics/mg_assembly/commit/db6eb89c97af100893bfab8a9cb6d90e4b9aef21))
* viral/quantify paths for reference ([e1cb62c](https://github.com/3d-omics/mg_assembly/commit/e1cb62cb41628448ff948891aa042911d283747f))
* **viral:** dedupe's output is a fasta ([ab34e77](https://github.com/3d-omics/mg_assembly/commit/ab34e778bd5ed3347a88a1724cb01e0bc244feb4))
* **viral:** disable checkv filtering ([2d4bbc7](https://github.com/3d-omics/mg_assembly/commit/2d4bbc7727725c1c0e331511d9dab635211c99d6))
* **viral:** disentangle rules from viral and eukaryotes ([155a767](https://github.com/3d-omics/mg_assembly/commit/155a7672b85ee7441908b0559392d860c5b294cc))
* **viral:** lining ([dbb3793](https://github.com/3d-omics/mg_assembly/commit/dbb37935e67e701112830a638da0ec3dc7d5499a))
* **viral:** missing folder and wrong output ([b00711a](https://github.com/3d-omics/mg_assembly/commit/b00711a2ac95deefca93c32f4e4f0c92f4a39c0b))
* **viral:** remove checkv ([8b213f5](https://github.com/3d-omics/mg_assembly/commit/8b213f59c4ad192fad871d50896776783bf0fccf))
* **viral:** remove checkv from environment ([134dad9](https://github.com/3d-omics/mg_assembly/commit/134dad923044ea32cae6d3232cd569f989accec9))
* **viral:** remove genomad gff generation ([1142cc7](https://github.com/3d-omics/mg_assembly/commit/1142cc758f28bd4de6caa0995381878893d1d649))
* **viral:** snakefmt ([0a2a2b4](https://github.com/3d-omics/mg_assembly/commit/0a2a2b4488609da00d474c17a1be333ce66ae9c7))
* **viral:** sort environment ([0e76477](https://github.com/3d-omics/mg_assembly/commit/0e76477bd799d207dcebb0efe5951860e27226cf))
* **virify:** rename rules ([14ff827](https://github.com/3d-omics/mg_assembly/commit/14ff827d774edb3f4e139938c20cfa72066082fd))
* **virsorter:** input fasta ([d682aad](https://github.com/3d-omics/mg_assembly/commit/d682aadd23909d3149508f3d95e14a058758e98c))
* **viruses:** annotation was not being run by default ([d46ba55](https://github.com/3d-omics/mg_assembly/commit/d46ba55a6e3e9dd6010b5b801702b0ded3c891f0))
* wrong out_dir ([29e3654](https://github.com/3d-omics/mg_assembly/commit/29e3654d599fa907c6fbbfc8daec393d5b8df7c1))


### Performance Improvements

* **coverm:** compress everything ([07934fc](https://github.com/3d-omics/mg_assembly/commit/07934fcfc319521b83062fe865ca2936fcef4c59))
* **dram:** half of parallel jobs, but using 2 threads ([fa6de49](https://github.com/3d-omics/mg_assembly/commit/fa6de49b5748bb6b8042046efa8f8adfc6d37a67))
* **dram:** use 64GB to be sure ([44b2a8b](https://github.com/3d-omics/mg_assembly/commit/44b2a8b141c6e9e67fe4f20c3c28d832558f0615))
* **genomad:** raise threads to 24 ([745459e](https://github.com/3d-omics/mg_assembly/commit/745459e10d7ce24a8c5e48f6c3e8c9a743d85f40))
* **gtdbtk:** start with 64GB ([f93e400](https://github.com/3d-omics/mg_assembly/commit/f93e400ee479ef54e0126abfeecc4f2dcc100b78))
* **nonpareil:** compress output ([4fd351b](https://github.com/3d-omics/mg_assembly/commit/4fd351b1c39786a04a7148f0f6b7af57274ed393))
* parallelize virsorter2 and dramv ([75fc9fc](https://github.com/3d-omics/mg_assembly/commit/75fc9fcabfe6127232de9000746e90e912e8c49f))
* raise dram-v threads ([8904943](https://github.com/3d-omics/mg_assembly/commit/89049438c3006cabb436b1ac6a8a9dbb70d30642))
* **singlem:** compress condense and otu table ([8948cba](https://github.com/3d-omics/mg_assembly/commit/8948cba24e107f72eea681b1f9dca28ee2f71dad))
* **singlem:** compress everything except jsons ([f891428](https://github.com/3d-omics/mg_assembly/commit/f89142874745ec81d46ae20fa0eb5e46eb983ab8))
* **singlem:** compress microbial fraction ([b203dc7](https://github.com/3d-omics/mg_assembly/commit/b203dc71f3973aebffea08392ad4660591a177b9))
* **viruses/annotate:** compress everything ([52f5c29](https://github.com/3d-omics/mg_assembly/commit/52f5c2934a179f8714ec1791802db2f53c98fefe))
